validator_client/src/duties_service.rs

```diff
@@ -674,84 +682,69 @@ async fn poll_beacon_attesters_for_epoch<T: SlotClock + 'static, E: EthSpec>(
         &[metrics::UPDATE_ATTESTERS_FETCH],
     );
 
-    let response = duties_service
-        .beacon_nodes
-        .first_success(
-            duties_service.require_synced,
-            OfflineOnFailure::Yes,
-            |beacon_node| async move {
-                let _timer = metrics::start_timer_vec(
-                    &metrics::DUTIES_SERVICE_TIMES,
-                    &[metrics::ATTESTER_DUTIES_HTTP_POST],
-                );
-                beacon_node
-                    .post_validator_duties_attester(epoch, local_indices)
-                    .await
-            },
-        )
-        .await
-        .map_err(|e| Error::FailedToDownloadAttesters(e.to_string()))?;
-
-    drop(fetch_timer);
-    let _store_timer = metrics::start_timer_vec(
-        &metrics::DUTIES_SERVICE_TIMES,
-        &[metrics::UPDATE_ATTESTERS_STORE],
-    );
+    // Request duties for either `INITIAL_DUTIES_QUERY_SIZE` validators or the count of validators for which we
+    // don't already know their duties for that epoch, whichever subset is larger. We use the `dependent_root`
+    // in the response to determine whether validator duties need to be updated. This is to ensure that we don't
+    // request for extra data unless necessary in order to save on network bandwidth.
+    let uninitialized_validators =
+        get_uninitialized_validators(duties_service, &epoch, local_pubkeys);
+    let indices_to_request = if !uninitialized_validators.is_empty() {
+        uninitialized_validators.as_slice()
+    } else {
+        &local_indices[0..min(INITIAL_DUTIES_QUERY_SIZE, local_indices.len())]
+    };
 
+    let response =
+        post_validator_duties_attester(duties_service, epoch, indices_to_request).await?;
     let dependent_root = response.dependent_root;
 
-    // Filter any duties that are not relevant or already known.
-    let new_duties = {
+    // Find any validators which have conflicting (epoch, dependent_root) values or missing duties for the epoch.
+    let validators_to_update: Vec<_> = {
         // Avoid holding the read-lock for any longer than required.
         let attesters = duties_service.attesters.read();
-        response
+        local_pubkeys
+            .iter()
+            .filter(|pubkey| {
+                attesters.get(pubkey).map_or(true, |duties| {
+                    duties
+                        .get(&epoch)
+                        .map_or(true, |(prior, _)| *prior != dependent_root)
+                })
+            })
+            .collect::<Vec<_>>()
+    };
+
+    if validators_to_update.is_empty() {
+        // No validators have conflicting (epoch, dependent_root) values or missing duties for the epoch.
+        return Ok(());
```

Comment:
I think we need to make sure we run `update_per_validator_duty_metrics` so that we start reporting the next-epoch duties as the `current_slot` progresses past the current-epoch slot.

Since we've changed `update_per_validator_duty_metrics`, perhaps it makes sense to hoist it up into `poll_beacon_attesters` (perhaps after each call to `poll_beacon_attesters_for_epoch` in [here](https://github.com/sigp/lighthouse/blob/cfaab2dcb93fe9790bf448bc0f253b51eebdd457/validator_client/src/duties_service.rs#L536-L578))? That way we can exit this whenever we like and still be confident that `update_per_validator_duty_metrics` is being called.

Reply:
Nice catch 🙏 , I've moved this to where you suggested!